pkg/cfn/manager/tasks_test.go

```diff
@@ -1,104 +1,328 @@
 
-				cc = newClusterConfig(clusterName)
+					subTask2 := &TaskTree{Parallel: false, Sub: true}
+					subTask2.Append(&taskWithoutParams{
+						info: "t2.1",
+					})
+					subTask3 := &TaskTree{Parallel: true, Sub: true}
+					subTask3.Append(&taskWithoutParams{
+						info: "t3.1",
+					})
+					subTask3.Append(&taskWithoutParams{
+						info: "t3.2",
+					})
+					tasks.Append(subTask2)
+					subTask1.Append(subTask3)
+
+					Expect(tasks.Describe()).To(Equal("2 sequential tasks: { 2 sequential sub-tasks: { t1.1, 2 parallel sub-tasks: { t3.1, t3.2 } }, t2.1 }"))
+				}
+			})
 
-				sc = NewStackCollection(p, cc)
+			It("should execute orderly", func() {
+				{
+					called11 := false
 
-				sucecssfulData = []string{}
+					tasks := &TaskTree{Parallel: false}
+					subTask1 := &TaskTree{Parallel: false, Sub: true}
+					subTask1.Append(&taskWithoutParams{
+						info: "t1.1",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(100 * time.Millisecond)
+								called11 = true
+								errs <- nil
+								close(errs)
+							}()
+							return nil
+						},
+					})
+					tasks.Append(subTask1)
 
-				call = func(errs chan error, data interface{}) error {
-					s := data.(string)
-					if s == "fail" {
-						return errors.New("call failed")
-					}
+					called21 := false
 
-					go func() {
-						defer close(errs)
+					subTask2 := &TaskTree{Parallel: false, Sub: true}
+					subTask2.Append(&taskWithoutParams{
+						info: "t2.1",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(150 * time.Millisecond)
+								called21 = true
+								errs <- fmt.Errorf("t2.1 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+					tasks.Append(subTask2)
 
-						sucecssfulData = append(sucecssfulData, s)
+					called31 := false
+					called32 := false
 
-						errs <- nil
-					}()
+					subTask3 := &TaskTree{Parallel: true, Sub: true}
+					subTask3.Append(&taskWithoutParams{
+						info: "t3.1",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(200 * time.Millisecond)
+								called31 = true
+								errs <- nil
+								close(errs)
+							}()
+							return nil
+						},
+					})
+					subTask3.Append(&taskWithoutParams{
+						info: "t3.2",
+						call: func(errs chan error) error {
+							go func() {
+								time.Sleep(350 * time.Millisecond)
+								called32 = true
+								errs <- fmt.Errorf("t3.2 always fails")
+								close(errs)
+							}()
+							return nil
+						},
+					})
+					subTask1.Append(subTask3)
 
-					return nil
+					Expect(tasks.Describe()).To(Equal("2 sequential tasks: { 2 sequential sub-tasks: { t1.1, 2 parallel sub-tasks: { t3.1, t3.2 } }, t2.1 }"))
+
+					errs := tasks.DoAllSync()
+					Expect(errs).To(HaveLen(2))
+					Expect(errs[0].Error()).To(Equal("t3.2 always fails"))
+					Expect(errs[1].Error()).To(Equal("t2.1 always fails"))
+
+					Expect(called11).To(BeTrue())
```

Comment:
This checks if tasks have been called, but it'd be nice to test the effective order of completion... I was wondering would be a good way to test that? Perhaps a channel or thread-safe slice with a mutex?

Reply:
Tried a channel, and it didn't work too well. Switched to a slice with a mutex, seems to work fine.